### DISCUSSION TITLE
NativeControlHost.CreateNativeControlCore() Doesn't Implement IMacOSTopLevelPlatformHandle (Native OpenGL)

### DISCUSSION THREAD

**Discussion by djrecipe** [AvaloniaUI/Avalonia#14495] [Q&A] [answered]
Hello all, I am trying to get the `NSView` from `NativeControlHost.CreateNativeControlCore()`.

However, it seems that `IPlatformHandle` here implements `PlatformHandle` but it doesn't implement `IMacOSTopLevelPlatformHandle`

How can I get the `NSView` so I can make native OpenGL calls?

**Comment by djrecipe** [comment]
Ultimately I'd like to add an `NSOpenGLView` as a subview to the view returned by `CreateNativeControlCore(parent)`

**Comment by BAndysc** [comment, marked as answer]
I had a similar use case and that was enough to attach a nsopenglview to a CreateNativeControlCore:

```
    protected override IPlatformHandle CreateNativeControlCore(IPlatformHandle parent)
    {
        uint[] attribs = new uint[]
        {
            AppleOpenGlConstants.NSOpenGLPFANoRecovery,
            AppleOpenGlConstants.NSOpenGLPFAAccelerated,
            AppleOpenGlConstants.NSOpenGLPFADoubleBuffer,
            AppleOpenGlConstants.NSOpenGLPFAColorSize, 24,
            AppleOpenGlConstants.NSOpenGLPFADepthSize, 24,
            AppleOpenGlConstants.NSOpenGLPFAOpenGLProfile,
            AppleOpenGlConstants.NSOpenGLProfileVersion4_1Core,
            0
        };
        ObjectiveC.NSOpenGLPixelFormat pixelFormat = new ObjectiveC.NSOpenGLPixelFormat(attribs);
        ObjectiveC.NSOpenGLContext context = new ObjectiveC.NSOpenGLContext(pixelFormat, ObjectiveC.Object.Null)
        {
        };    
        var nsView = new ObjectiveC.NSOpenGLView(new ObjectiveC.NSRect(0, 0, 100, 100), pixelFormat)
        {
            WantsBestResolutionOpenGLSurface = true
        };
        context.View = nsView;
        nsView.OpenGlContext = context;
        return new PlatformHandle(nsView.Pointer, "HWND"); 
    }
```

you can see all the code here: https://github.com/BAndysc/WoWDatabaseEditor/blob/443f0aaace8e27230ec95f1375864bb071d0173c/Rendering/TheAvaloniaOpenGL/NativeOpenGlControlBase.cs#L166

**Reply by djrecipe** [reply]
Thanks boss, this has put me on the right track.

Although it seems that no matter what I do, I get an error claiming that I'm not on the UI thread. Luckily we can just set `NSApplication.CheckForIllegalCrossThreadCalls = false;`

Once I've made a bit more progress then I'll post my solution as well, as it seems my namespaces deviate slightly from what you have suggested.
